docs/in-depth-guide-ingredient-groups.md

```diff
@@ -0,0 +1,125 @@
+
+* https://cooking.nytimes.com/recipes/1024570-green-salad-with-warm-goat-cheese-salade-de-chevre-chaud
+* https://lifestyleofafoodie.com/air-fryer-frozen-french-fries
+* https://www.bbcgoodfood.com/recipes/monster-cupcakes
+
+Not all websites use ingredient groups, and those that do use ingredient groups will not use them for all recipes.
+
+This library allows a scraper to return the ingredients in the groups defined in the recipe, if the functionality is added to that scraper.
+
+## `ingredient_groups()`
+
+The `ingredient_groups()` function returns a list of `IngredientGroup` objects. Each `IngredientGroup` is a dataclass that represents a group of ingredients:
+
+```python
+@dataclass
+class IngredientGroup:
+    ingredients: List[str]
+    purpose: Optional[
+        str
+    ] = None  # this group of ingredients is {purpose} (e.g. "For the dressing")
+```
+
+The *purpose* is the ingredient group heading, such as *"For the dressing"*, *"For the sauce"* etc. The *ingredients* is the list of ingredients that comprise that group.
+
+This dataclass is defined in `_grouping_utils.py` and should be imported from there
+
+```pyth
+from ._grouping_utils import IngredientGroup
+```
+
+The `ingredient_groups()` function has a default implementation in the `AbstractScraper` class that returns a single `IngredientGroup` object with `purpose` set to `None` and `ingredients` set to the output from the scraper's `ingredients()` function.
+
+To add ingredient group support to a scraper, the `ingredient_group` function needs to be overridden in the scraper class. The are three main things to consider when implementing ingredient groups support:
+
+1. The groups and their contents are not present in the Recipe schema. They must be extracted from the recipe HTML.
+2. The ingredients found in `ingredients()` and `ingredient_groups()` must be the same. This may sound obvious but there can sometimes minor differences in the ingredients in the schema and the ingredients in the HTML.
+3. Not all recipes on a website will use ingredient groups, so the implementation must support recipes that do and recipes that don't have ingredient groups. For recipes that don't have ingredient groups, the output should be the same as default implementation (i.e. a single `IngredientGroup` with `purpose=None` and `ingredients=ingredients()`).
+
+In many cases the structure of how ingredients and group heading appear in the HTML is very similar. Some helper functions have been developed to make the implementation easier.
+
+## _grouping_utils.py
+
+The `_grouping_utils.py` file contains a helper function (`group_ingredients(...)`) that will handle the extraction of ingredients groups and their headings from the HTML, make sure the ingredients in the groups match those return from `.ingredients()`, and then return the groups.
+
+The `group_ingredients()` function takes four arguments:
+
+```pyth
+def group_ingredients(
+    ingredients_list: List[str],
+    soup: BeautifulSoup,
```

Comment:
Not directly related to this documentation: I've wondered before whether we should relocate `group_ingredients` onto `AbstractScraper` so that we don't have to pass the `soup` argument here.

Maybe if-and-when that is refactored, that'd be a good case for updating documentation and codebase at the same time.

Reply:
I had thought along on the same lines when implementing `group_ingredients`. It felt like adding a lot to the `AbstractScraper` that most scrapers probably wouldn't use.

Reply:
Yep, seems fair, and it's a small detail.  We can skip that and maybe reconsider it later :+1: